libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/targets/jvm/KotlinWithJavaTarget.kt

```diff
@@ -44,10 +45,9 @@ open class KotlinWithJavaTarget<KotlinOptionsType : KotlinCommonOptions>(
             KotlinWithJavaCompilationFactory(project, this, kotlinOptionsFactory)
         )
 
-    internal val defaultArtifactClassesListFile: File
-        get() {
+    internal val defaultArtifactClassesListFile: Provider<File> = project.provider {
             val jarTask = project.tasks.getByName(artifactsTaskName) as Jar
-            return File(File(project.buildDir, KOTLIN_BUILD_DIR_NAME), "${sanitizeFileName(jarTask.archiveFileName.get())}-classes.txt")
+            File(File(project.buildDir, KOTLIN_BUILD_DIR_NAME), "${sanitizeFileName(jarTask.archiveFileName.get())}-classes.txt")
         }
```

Comment:
❓ I'm not sure where this `val` is wired, but the above code is losing task dependency and reference to `jar.archiveFileName` and could cause ordering problems.

Another way to write this would be:

```kotlin
project.tasks.named("artifactsTaskName", Jar::class).flatMap { jar ->
    project.layout.buildDirectory.file("$KOTLIN_BUILD_DIR_NAME/${sanitizeFileName(jar.archiveFileName.get())}-classes.txt")
}
```
This would return a `Provider<RegularFile>`. If you need it to be a `Provider<File>` just chain with `.map { it.asFile }`

Again, I'm not sure where this gets used and if my comment is relevant.